src/hotspot/os/linux/os_linux.cpp

```diff
@@ -4226,6 +4226,90 @@ char* os::pd_attempt_reserve_memory_at(char* requested_addr, size_t bytes, bool
   return nullptr;
 }
 
+char* os::pd_attempt_reserve_memory_below(char* max, size_t bytes, size_t alignment,
+                                          int max_attempts) {
+
+  assert(is_aligned(alignment, os::vm_allocation_granularity()), "alignment unaligned");
+  assert(is_aligned(bytes, os::vm_page_size()), "size unaligned");
+
+  // Scan /proc/self/maps for an address space hole that is suitable for our reservation. If found,
+  // attempt to map into that hole.
+
+  log_trace(os, map)("attempt_reserve_memory_below " PTR_FORMAT ", bytes: " SIZE_FORMAT_X
+                     ", alignment: " SIZE_FORMAT_X, p2i(max), bytes, alignment);
+
+// somewhat shorter
+#define ALGNUP(x) align_up(x, alignment)
+#define ALGNDWN(x) align_down(x, alignment)
+
+  FILE* f = os::fopen("/proc/self/maps", "r");
```

Comment:
How sane is it to parse /proc stuff, especially when done line-by-line? Is this not racy? I am not familiar with these things. I assume there is no sane(r) API for the purpose?